lib/MatrixGitterBridge.js

```diff
@@ -0,0 +1,302 @@
+          }
+          else if (membership == "join") {
+            self.onBotJoined(event.room_id);
+          }
+          else if (membership == "leave") {
+            // TODO: "leave" events might mean we left the room, or got kicked
+            //   while still in invite state - i.e. invite was cancelled
+            //   before we joined.
+            self.onBotLeft(event.room_id);
+          }
+          else {
+            console.log('matrix member event for myself to state ' + membership + 
+                        ' from state ', event);
+          }
+          return;
+        }
+
+        if (event.type !== "m.room.message" || !event.content) {
+          return;
+        }
+
+        if (event.sender == bridge.getBot().getUserId()) {
+          return;
+        }
+
+        console.log('matrix->' + event.room_id + ' from ' + event.user_id + ':', event.content.body)
+
+        var handled = false;
+
+        if (config.matrix_admin_room && event.room_id == config.matrix_admin_room) {
+          self.onMatrixAdminMessage(event);
+          handled = true;
+        }
+
+        var bridgedRoom = self._bridgedRoomsByMatrixId[event.room_id];
+        if (bridgedRoom) {
+          bridgedRoom.onMatrixMessage(event);
+          handled = true;
+        }
+
+        if (!handled) {
+          console.log("  Wasn't expecting this room; ignore");
+        }
+      }
+    }
+  });
+
+  this._bridge = bridge;
+  this._gitterUserId = null;
+
+  this._gitterusers = {};
+
+  var rules = [];
+  if (config.name_mangling) {
+    for (var i = 0; i < config.name_mangling.length; i++) {
+      var rule = config.name_mangling[i];
+
+      rules.push({
+        pattern: new RegExp(rule.pattern),
+        template: rule.template
+      });
+    }
+  }
+  this._name_mangling_rules = rules;
+}
+
+MatrixGitterBridge.prototype.getMyGitterUserId = function() {
+  if (this._gitterUserId)
+    return Promise.resolve(this._gitterUserId);
+
+  return this._gitter.currentUser().then((u) => {
+    this._gitterUserId = u.id;
+    return u.id;
+  });
+};
+
+MatrixGitterBridge.prototype.mangleName = function(name) {
+  var rules = this._name_mangling_rules;
+
+  for (var i = 0; i < rules.length; i++) {
+    var rule = rules[i];
+    var matches = rule.pattern.exec(name);
+    if (!matches)
+      continue;
+
+    // TODO: more groups?
+    return rule.template.replace('$1', matches[1]);
+  }
+
+  return name;
+}
+
+MatrixGitterBridge.prototype.mapGitterUser = function(user) {
+  var id = user.id;
+
+  return this._gitterusers[id] = (this._gitterusers[id] ||
+      this.getRemoteUser(id).then((model) => {
+        if(!model)
+          model = new RemoteUser(id, {username: user.username});
+
+        return new GitterUser(this, model);
+      }));
+};
+
+MatrixGitterBridge.prototype.getGitterUserById = function(id) {
+  var user = this._gitterusers[id];
+  if (user)
+    return user;
+
+  return this.getRemoteUser(id).then((model) => {
```

Comment:
If you call `getGitterUserById` quickly in a row, the guard on `:152` will not protect this function from being called multiple times. This is practically fine as it stands as it's just doing a DB lookup, but you should be aware of this (and protect it if you feel you need to).